// @ngneat/hot-toast/esm2022/lib/hot-toast.module.mjs
import { NgModule, makeEnvironmentProviders } from '@angular/core';
import { ToastConfig } from './hot-toast.model';
import * as i0 from "@angular/core";
/**
 * @deprecated Use `provideHotToastConfig` to provide a global toast configuration.
 */
export class HotToastModule {
    static forRoot(config) {
        return {
            ngModule: HotToastModule,
            providers: [{ provide: ToastConfig, useValue: config }],
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.1", ngImport: i0, type: HotToastModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.1", ngImport: i0, type: HotToastModule }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.1", ngImport: i0, type: HotToastModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.1", ngImport: i0, type: HotToastModule, decorators: [{
            type: NgModule
        }] });
export function provideHotToastConfig(config) {
    return makeEnvironmentProviders([{ provide: ToastConfig, useValue: config }]);
}
